pcap-etw.c

```diff
@@ -0,0 +1,397 @@
+	
+	EVENT_TRACE_LOGFILE evt_config;
+	ZeroMemory(&evt_config, sizeof(EVENT_TRACE_LOGFILE));
+	evt_config.LoggerName = etw->session_name;
+	evt_config.LogFileName = NULL;	// indicates a real-time session rather than a logfile
+	evt_config.ProcessTraceMode = PROCESS_TRACE_MODE_EVENT_RECORD | PROCESS_TRACE_MODE_REAL_TIME;
+	evt_config.EventRecordCallback = pcap_etw_process_event;
+	evt_config.BufferCallback = pcap_etw_buffer_event;
+	evt_config.Context = etw; // Etw context
+
+	// Get a handle to the trace session:
+	etw->trace = OpenTrace(&evt_config);
+	if (INVALID_PROCESSTRACE_HANDLE == etw->trace)
+	{
+		pcap_snprintf(self->errbuf, PCAP_ERRBUF_SIZE, "Unable to open trace");
+		return (-1);
+	}
+
+	etw->thread = CreateThread(NULL, 0,	pcap_etw_listener, (LPVOID)etw->trace, 0, NULL);
+	if (INVALID_HANDLE_VALUE == etw->thread)
+	{
+		pcap_snprintf(self->errbuf, PCAP_ERRBUF_SIZE, "Unable to create a thread listener");
+		return (-1);
+	}
+
+	return (0);
+}
+
+/*
+ *	Use the software base filtering engine
+ */
+static int
+pcap_setfilter_op_etw(pcap_t *p, struct bpf_program *fp)
+{
+	return install_bpf_program(p, fp);
+}
+
+/*
+ *	Check stat for this layer capture
+ */
+static int
+pcap_stats_etw(pcap_t *p, struct pcap_stat *ps)
+{
+	ps->ps_drop = 0;
+	//ps->ps_recv = 0;
+	return (0);
+}
+
+/*
+ *	Read operation pending list event from the recorded list
+ */
+static int
+pcap_read_op_etw(pcap_t *p, int cnt, pcap_handler callback, u_char *user)
+{
+	pcap_etw_t* etw = (pcap_etw_t*)p->priv;
+	int n = 0;
+	for (n = 0; n < cnt; n++)
+	{
+		pcap_etw_list_event_t* element = NULL;
+		// no more elements
+		while ((element = pcap_etw_list_pop(&etw->events)) == NULL)
+		{
+			DWORD status = WaitForSingleObject(etw->events.wait, 1000);
+			switch (status)
+			{
+			case WAIT_OBJECT_0:
+				continue;
+			case WAIT_TIMEOUT:
+				return (n);
+			case WAIT_FAILED:
+				return (-1);
+			}
+			
+		}
+
+		struct pcap_pkthdr header;
+		header.caplen = element->event->Size;
+		header.len = element->event->Size;
+
+		/* convert between timestamp formats */
+		ULONGLONG ts = element->event->TimeStamp.QuadPart;
+		header.ts.tv_sec = (int)(ts >> 32);
+		ts = (ts & 0xffffffffi64) * 1000000;
+		ts += 0x80000000; /* rounding */
+		header.ts.tv_usec = (int)(ts >> 32);
+		if (header.ts.tv_usec >= 1000000) {
+			header.ts.tv_usec -= 1000000;
+			header.ts.tv_sec++;
+		}
+
+		p->stat.ps_recv++;
+
+		callback(user, &header, (u_char*)element->event);
+
+		free(element->event);
+		free(element);
+	}
+	
+	return (n);
+}
+
+/*
+ *	Create new etw interface
+ */
+pcap_t *
+pcap_create_interface(const char *device _U_, char *ebuf)
+{
+	pcap_t *handle = pcap_create_common(ebuf, sizeof(pcap_etw_t));
+	if (handle == NULL)
+		return NULL;
+
+	handle->activate_op = pcap_activate_op_etw;
+	handle->setfilter_op = pcap_setfilter_op_etw;
+	handle->stats_op = pcap_stats_etw;
+	handle->read_op = pcap_read_op_etw;
+	handle->snapshot = MAXIMUM_SNAPLEN;
+	handle->linktype = DLT_USER0;
```

Comment:
Don't use a DLT_USER value; send an email to tcpdump-workers@lists.tcpdump.org to request that a DLT_ value be assigned for this purpose.

Reply:
I'm just sending  the email. Thanks for your feedback.